features/lorawan/LoRaWANStack.cpp

```diff
@@ -492,322 +363,526 @@ int16_t LoRaWANStack::handle_rx(uint8_t* data, uint16_t length, uint8_t& port, i
+    _device_current_state = DEVICE_STATE_RECEIVING;
+
+    _loramac.on_radio_rx_done();
+
+    loramac_mhdr_t mac_hdr;
+    uint8_t pos = 0;
+    mac_hdr.value = payload[pos++];
+
+    /**
+     * Handle a Join Accept frame
+     */
+    if (mac_hdr.bits.mtype == FRAME_TYPE_JOIN_ACCEPT) {
+        if (_loramac.nwk_joined()) {
+            _loramac.abort_rx();
+            _ready_for_rx = true;
+            return;
+        }
+
+        _loramac.handle_join_accept_frame(payload, size);
+
+        if (_loramac.get_mlme_confirmation()->status == LORAMAC_EVENT_INFO_STATUS_OK) {
+            state_controller(DEVICE_STATE_CONNECTED);
+            _ready_for_rx = true;
+            return;
+        }
+
+        tr_error("Malformed JOIN_ACCEPT_FRAME");
+        _ready_for_rx = true;
+        return;
     }
 
-    _loramac.setup_link_check_request();
-    return LORAWAN_STATUS_OK;
-}
+    /**
+     * Handle data frames
+     */
+    _ctrl_flags &= ~LORAMAC_MSG_RECVD;
+    _loramac.handle_data_frame(payload, size, pos, mac_hdr.bits.mtype, rssi, snr);
 
-void LoRaWANStack::remove_link_check_request()
-{
-    _link_check_requested = false;
+    if (_loramac.get_mlme_confirmation()->pending) {
+        _loramac.post_process_mlme_request();
+        mlme_confirm_handler();
+    }
+
+    if (_loramac.get_mcps_indication()->type == MCPS_CONFIRMED) {
+        // if ack was not received, we will try retransmission after
+        // ACK_TIMEOUT. handle_data_frame() already disables ACK_TIMEOUT timer
+        // if ack was received
+        if (_loramac.get_mcps_indication()->is_ack_recvd) {
+            tr_debug("Ack=OK, NbTrials=%d", _loramac.get_mcps_confirmation()->nb_retries);
+            _loramac.post_process_mcps_req();
+            _ctrl_flags |= LORAMAC_TX_DONE;
+            state_controller(DEVICE_STATE_STATUS_CHECK);
+        }
+    } else {
+        // handle UNCONFIRMED case here, RX slots were turned off due to
+        // valid packet reception
+        _loramac.post_process_mcps_req();
+        _ctrl_flags |= LORAMAC_TX_DONE;
+        state_controller(DEVICE_STATE_STATUS_CHECK);
+    }
+
+    // handle any pending MCPS indication
+    if (_loramac.get_mcps_indication()->pending) {
+        _loramac.post_process_mcps_ind();
+        _ctrl_flags |= LORAMAC_MSG_RECVD;
+        state_controller(DEVICE_STATE_STATUS_CHECK);
+    }
+
+    // change the state only if a TX cycle completes for Class A
+    // For class C it's not needed as it will already be in receiving
+    // state, no matter if the TX cycle completed or not.
+    if (!(_ctrl_flags & LORAMAC_TX_ONGOING)) {
+        // we are done here, update the state
+          state_machine_run_to_completion();
+    }
+
+    if (_loramac.get_mlme_indication()->pending) {
+        tr_debug("MLME Indication pending");
+        _loramac.post_process_mlme_ind();
+        tr_debug("Automatic uplink requested");
+        mlme_indication_handler();
+    }
+
+    memset(_rx_payload, 0, sizeof _rx_payload);
```

Comment:
Why the memset? memsets are one of the most expensive things you can do - you don't want this sort of thing in data paths. Although I guess this isn't very high throughput...